content/blog/can-we-teach-software-design/index.md

```diff
@@ -0,0 +1,65 @@
+---
+title: "Can we teach software design? A drummer's perspective"
+date: "2022-04-29"
+description: "Software design is somewhat abstract and difficult to teach. Learning to play the drums is somewhat similar - can we draw some parallels and find a better path forward?"
+published: false
+tags: ["software-design"]
+---
+Considering how much software design[^1] is part of a dev's daily work, the availability of good, clear resources seems disproportionate. There's a lot of material on using X library or Y framework, but whenever I've asked or looked up software design,  in general I get some disjointed links to patterns or principles, along with advice that "it comes with experience".
+
+[^1]: Software Design is itself a broad term (see [wikipedia](https://en.wikipedia.org/wiki/Software_design)). Here I think more about "code-level" aspects e.g. how to decompose a problem into different modules, create extensible code. But the thinking could also be applied at "higher-levels" e.g. breaking out microservices.
+
+While there's truth to that advice, I wonder if we could do better. Maybe this is difficult because software design is a pursuit that is relatively abstract and difficult to measure accurately. In that way, it's somewhat similar to creative pursuits. I used to play the drums, and when asking drummers to explain why their idols are so great, it's common to summarise as "they have great feel" - sound familiar?
+
+So, can we draw some parallels between learning to play the drums and software design? Could it improve how we frame and approach the teaching of software design?
+
+## How Drummers Practice
+I'd broadly categorise most practice as achieving one of two main goals: "how to play", and "when to play"[^2].
+
+[^2]: In reality, some practice exercises straddle both buckets. There are also other aspects such as being able to sight-read drum score, but I feel they mainly support learning than being an end themselves.
+
+### How to Play
+This is mainly about mechanics - building a library of motions (e.g. grooves, fills) and getting comfortable with executing them (e.g. technique, building muscle memory, playing up to speed). These "building blocks" can then be played in sequence to a song.
+
+### When to Play
+This is about knowing the "right" sequence of building blocks to play e.g. does the groove fit this song? Is this the right time to play a particular fill? What sort of fill should I play?
+
+There are principles and even some heuristics, such as:
+- "play a groove for 3 bars, do a fill on the 4th"
```

Comment:
Would love to see a video descriptor of a groove here 😄 This sounds so cool! Maybe an embed?